### PR TITLE
8319451: PhaseIdealLoop::conditional_move is too conservative

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

When transforming a Phi into a CMove, the threshold is set to be approximately BlockLayoutMinDiamondPercentage, the reason is given:

    // BlockLayoutByFrequency optimization moves infrequent branch
    // from hot path. No point in CMOV'ing in such case

This sets the default value of the threshold to be around 18%, which is too conservative. The reason also does not make a lot of sense since the important property which makes jumping expensive is not code layout. We should remove this.

Please kindly review, thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8319451](https://bugs.openjdk.org/browse/JDK-8319451): PhaseIdealLoop::conditional_move is too conservative (**Enhancement** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [4513cbef](https://git.openjdk.org/jdk/pull/16524/files/4513cbef699e62144d5177c358ffff48b6c888ea)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [4513cbef](https://git.openjdk.org/jdk/pull/16524/files/4513cbef699e62144d5177c358ffff48b6c888ea)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16524/head:pull/16524` \
`$ git checkout pull/16524`

Update a local copy of the PR: \
`$ git checkout pull/16524` \
`$ git pull https://git.openjdk.org/jdk.git pull/16524/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16524`

View PR using the GUI difftool: \
`$ git pr show -t 16524`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16524.diff">https://git.openjdk.org/jdk/pull/16524.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16524#issuecomment-1796085428)

### CODE REVIEW

**Comment by merykitty** [comment]
This is the result of the benchmark running on my machine. The result is consistent with the common knowledge, as the `CMove` version seems to offer better performance when the uncommon path frequency is above 1%.

                                        Before             After
    Benchmark  (freq)  Mode  Cnt    Score     Error    Score    Error  Units
    CMove.run       3  avgt    5  179.402 ±   5.504  181.112 ±  8.658  us/op
    CMove.run       6  avgt    5  191.884 ±  15.271  189.571 ± 11.785  us/op
    CMove.run      10  avgt    5  196.144 ±   8.946  198.864 ±  1.440  us/op
    CMove.run      20  avgt    5  287.090 ± 141.540  198.979 ±  4.015  us/op
    CMove.run      30  avgt    5  353.721 ± 109.473  198.730 ±  2.302  us/op
    CMove.run      60  avgt    5  489.541 ±   4.100  199.645 ±  4.552  us/op
    CMove.run     100  avgt    5  675.543 ±  12.810  199.739 ±  2.202  us/op
    CMove.run     200  avgt    5  199.176 ±   2.957  199.584 ±  2.111  us/op
    CMove.run     300  avgt    5  198.941 ±   2.642  200.112 ±  3.551  us/op
    CMove.run     600  avgt    5  199.266 ±   3.999  199.023 ±  1.202  us/op

File: test/micro/org/openjdk/bench/vm/compiler/CMove.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Thread)
+@Warmup(iterations = 5, time = 1)
+@Measurement(iterations = 5, time = 1)
+@Fork(value = 1)
+public class CMove {
+    static final int SIZE = 1000000;
+
+    @Param({"3", "6", "10", "20", "30", "60", "100", "200", "300", "600"})
+    int freq;
```
Comment (cl4es) [inline test/micro/org/openjdk/bench/vm/compiler/CMove.java:40]: That `freq` is expressed in "occurrences per thousand" was only obvious after reading the code - perhaps a probability between 0 and 1 (with the appropriate adjustment to `r.nextFloat() < freq` below) would be slightly more intuitive?
```suggestion
    @Param({"0.003", "0.006", "0.01", "0.02", "0.03", "0.06", "0.1", "0.2", "0.3", "0.6"})
    float freq;
```

**Review by cl4es — COMMENTED** [review]
Nice, especially the property that run-to-run variance drops in the 1-20% range. 

@TobiHartmann told me he'll review and that he has queued up a batch of benchmarks to evaluate this change more widely.

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me. All tests passed and performance results look neutral (no statistically significant improvements or regressions).

**Review by vnkozlov — COMMENTED** [review]
Looks fine to me.

Looking on history of this code and I added it to address [JDK-7097546](https://bugs.openjdk.org/browse/JDK-7097546).

But later it was found not correct for some case and I even had similar fix prototype: [JDK-8034833](https://bugs.openjdk.org/browse/JDK-8034833). There was additional changes proposed there: in `block.hpp` and `.ad` file.

Please, look on attached in that report test and additional code changes there. May be be we can improve more `cmove`.  It could be done separately from this your fix if you want to spend more time on it.

**Comment by merykitty** [comment]
@vnkozlov I have investigated a little bit.

For these kinds of loops

    public static int test(int result, int limit, int mask) { // mask = 15
        for (int i = 0; i < limit; i++) {
          if ((i&mask) == 0) result++; // Non frequent
        }
        return result;
    }

Since this loop is perfectly predictable, no threshold of `CMove` transformation may offer performance advantages. I don't think this predictable branch is common, though.

Regarding the register pressure relating to a `CMove`, the main issue is that our local code motion does not do much (it does some heuristics around calls; the other element is block-wise latency, which is kind of useless in LCM context), I have tried some heuristics but it is easy to find a case where it is insufficient. I think it is probably a good idea to reimplement LCM using a more optimal algorithm.

**Review by vnkozlov — APPROVED** [review]
Thank you for additional investigation.

File: test/micro/org/openjdk/bench/vm/compiler/CMove.java (hunk 1 above)
Comment (merykitty) [reply test/micro/org/openjdk/bench/vm/compiler/CMove.java]: @cl4es You are right, I have fixed that.

**Comment by merykitty** [comment]
@TobiHartmann @vnkozlov @cl4es Thanks a lot for your reviews and testings, if there is nothing that concerns you, I will integrate the patch,

**Review by cl4es — APPROVED** [review]
Thank you for the work on this and for picking up on my suggestions for the microbenchmark! Looking forward to see the effects of this on a wider selection of benchmarks.

**Pull request closed without merging**